_posts/2018-01-14-smaller-docker-images.md

```diff
@@ -0,0 +1,405 @@
+```
+
+And you can compile the image as usual with:
+
+```bash
+$ docker build -t node-distroless .
+```
+
+The application should run as normal. To verify that is still the case, you could run the container like this:
+
+```bash
+$ docker run -p 3000:3000 node-distroless
+```
+
+And visit the page at [http://localhost:3000](http://localhost:3000).
+
+Is the image without all the extra binaries smaller?
+
+```bash
+$ docker images | grep node-distroless
+node-distroless                 latest              7b4db3b7f1e5        3 minutes ago       76.7MB
+```
+
+**That's only 76.7MB!**
+
+More than 600MB down from your previous image!
+
+This is extremely good news, but there's something you should pay attention to when it comes to distroless.
+
+When you container is running and you wish to inspect it, you can attach to a running container with:
+
+```bash
+$ docker exec -ti <inser_docker_id> bash
+```
+
+Attaching to a running container and running `bash` feels like establing an SSH session.
+
+But since distroless is a stripped down version of the original operating system, there are no extra binaries. There's no shell in the container!
+
+How can you attach to a running container if there's no shell?
+
+The good and the bad news is that you can't.
+
+It's bad news because you can only execute the binaries in the container. The only binary you could use is Node.js:
+
+```bash
+$ docker exec -ti <inser_docker_id> node
+```
+
+It's good news because an attacker exploiting your application and gaining access to the container won't be able to do as much damage as if were to access a shell. In other words, less binaries means smaller sizes and increased security. But at the cost of more painful debugging.
+
+But what if you cared about debugging and smaller sizes?
+
+## 3. Smaller base images with Alpine
+
+You could replace the distroless base image with an Alpine based image.
+
+Alpine Linux is:
+
+> a security-oriented, lightweight Linux distribution based on musl libc and busybox
+
+In other words, a linux distribution that is smaller in size.
+
+You shouldn't take their words for granted. Let's check if that's the case.
+
+Let's tweak the `Dockerfile` to use `node:8-alpine`:
+
+```dockerfile
+FROM node:8 as build
+
+WORKDIR /app
+COPY package.json .
+RUN yarn install
+COPY index.js .
+
+FROM node:8-alpine
+
+COPY --from=build /app /
+EXPOSE 3000
+CMD ["node", "index.js"]
+```
+
+You can build the image with:
+
+```bash
+$ docker build -t node-alpine .
+```
+
+And you can check the size with:
+
+```bash
+$ docker images | grep node-alpine
+node-alpine                     latest              aa1f85f8e724        38 seconds ago      69.7MB
+```
+
+**69.7MB!**
+
+Even smaller than the distroless image!
+
+Can you attach to a running container, unlike distroless? It's time to find out.
+
+Let's start the container first:
+
+```bash
+$ docker run -p 3000:3000 -d node-alpine
+9d8e97e307d705e7eb7cb714f16eccd19afabd5eea9db09917eeb8a4eeb36767
```

Comment:
do you really need to use the full hash, why not name --name=myfinecontainer the container so that people can copy paste the command?

Reply:
The full hash is what the command returns. If it is confusing I could remove it.

Reply:
maybe resist `-d` unless you're going to also describe stopping/killing you could also add `-ti --rm`

Reply:
But still I can't stop the container with Ctrl+C

Reply:
you can if you catch it in your JS
```js
process.on('SIGINT', process.exit);
```
 or `npm start` which will be pid `1` and the `node index.js` becomes pid `2` and `^C` works

Reply:
Or... magic `--init` flag. So awesome. Much magic. Very special.